### PR TITLE
Implement `Display` for `TypeId`s;  Add type parameters to the definition of custom types

### PR DESCRIPTION
**Pull request by sezna**
I've been working on unblocking vectors and allowing unused generics. I'm blocked on some changes in #1433 . For the sake of atomic PR's, I figured I'd wrap up the changes I've made thus far into a PR:

1. Include `type_parameters` in the definition of a struct or enum in the type engine. This allows for handling of generic types where the type is not necessarily used in the fields or variants.
2. Change `TypeId` from a type alias to a newtype. This provides greater type safety but also allowed me to implement `Display` for it. Now when you print a `TypeId`, instead of the numeric value in the type engine, you'll see the actual type it represents.
3. Do a bit of refactoring/code cleanup in places.

### CODE REVIEW

File: Cargo.lock [hunk 1]
```diff
@@ -102,9 +102,9 @@ dependencies = [
 
 [[package]]
 name = "anyhow"
-version = "1.0.56"
+version = "1.0.57"
```
Comment (adlerjohn) [inline Cargo.lock:105]: Why the change to lockfile?

**Review by adlerjohn — COMMENTED** [review]
Also, CI is failing

File: sway-core/src/type_engine.rs [hunk 2]
```diff
@@ -13,8 +16,33 @@ pub use type_info::*;
...
+pub struct TypeId(usize);
 
+impl std::ops::Deref for TypeId {
+    type Target = usize;
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl Display for TypeId {
```
Comment (mohammadfawaz) [inline sway-core/src/type_engine.rs:29]: image

File: Cargo.lock (hunk 1 above)
Comment (mohammadfawaz) [reply Cargo.lock:105]: So that's been happening to me sometimes and not sure why (especially when I'm cutting a release). I usually just unstage `Cargo.lock`, restore it, and then rebuild.

**Comment by mitchmindtree** [comment]
Might be particularly good for @emilyaherbert to review as there's potentially quite a bit of overlap in monomorphisation / declaration checking with #1433?

**Review by otrho — APPROVED** [review]
I like it.

**Review by mitchmindtree — APPROVED** [review]
> potentially quite a bit of overlap in monomorphisation / declaration checking with https://github.com/FuelLabs/sway/pull/1433?

I'm looking again and I think I must have been mistaken or looking at a different PR :woozy_face: The only real overlapping change is moving the struct declaration stuff, but that's easy enough to address. Lgtm too!
